### ISSUE TITLE
Implement a robust testing strategy

### ISSUE THREAD

**Issue by bobjansen** [eddelbuettel/anytime#26]
Results of calls to `anytime` depend on local settings which makes it a bit more tricky to write good tests, they can fail in different environments. Ideally we make sure that
1. Tests run correctly independent of the environment;
2. Still correctly simulate a wide array of locales inside 1 R session (no stop session, change setting, start session)

**Comment by eddelbuettel** [comment]
Well ... I am aware of the problem as you can tell from looking at `tests/`.  

What else can you suggest, concretely?

**Comment by bobjansen** [comment]
I've nothing concrete yet but I wanted to put it somewhere as an issue to be considered.

I have not tried any testing under different OS'es but I can spin up some VM's with different settings and start testing. Then I'll see how far I can get coding up some working tests with some locale and platform specific initialization.

**Comment by eddelbuettel** [comment]
I have been putting packages onto CRAN for maybe fifteen years. I have been using the available testing frameworks (via examples, then the `tests/` directory, often the [RUnit](https://cran.r-project.org/package=RUnit) package, was one of the folks driving code and adoption of [Travis CI](https://travis-ci.org) for R etc pp).   I think I have a reasonable handle on testing, and I don't see how what you suggest can be done in an automated fashion. 

Now, if you happen to have a lot of different local 'metal' and VMs then that is definitely something we can tap into.  I have a few packages that run tests _conditionally_.  

We also have a need for testing across different `TZ` values at on (maybe some versions of) Windows as there is still the weird Australia bug.

**Comment by bobjansen** [comment]
I know your track record, it is beyond dispute and I enjoy the fruits of your work on an almost daily basis. I just want to minimize the change that `anytime` subtly breaks without me or anyone else noticing.

I've probably misstated my intention: the idea is to create working tests on a number of platforms using one codebase through careful setup. When that's done I'd hope it will be relatively straightforward to add new tests for new functionality that isn't directly related to `TZ` and locale issues. Of course, we could disable (some) tests for stubborn platforms. I know it's not ideal and it might be quite a bit of work but I'm willing to give it a shot. At the very least we should be able to prevent accidents happening on Linux.

I own only a PC x86 box so I'm afraid I'm a bit limited there. I do have a big disk and a modicum of patience though so I can spin up plenty of VM's. 

Let me give it a try, and see what happens. We could move the issue to my repo?

**Comment by eddelbuettel** [comment]
I'm fine with closing this here, and coming back when we have more.

I too would like to test more, but it is so darn hard to automate this.  We (ideally) want different OSs and different releases and different timezones and ...

Now, I will admit that I am a little lazy^Hconservative in sticking to Travis CI and Linux.  One _can_ try appveyor there (but I am so disenchanted with Windows that I do not bother -- win-builder, and now  R Hub, works for me; ditto for OS X which I won't add on Travis CI for as long as I don't have a personal machine).   So we could have a shadow repo which frequently updates and runs those.  That would be an improvement.   And if you take care of that I don;t have to ;-)

Ideally we'd also test different TZ values and all that.  Some of that we should be able to do by software.

So in short: in favor of my experiments.  Not so much in favor of turning my (working !!) workflow upside down.

**Comment by eddelbuettel** [comment]
Closing this now as we have nothing immediately actionable. We agree on the idea and the need for this, the devil is making something automatically executable (beyond the basic testing frameworks we have in R).

**Comment by eddelbuettel** [comment]
I had an idea.  Look at how `Rscript tests/allFormats.R` in essence creates _N_ times the same string.

We didn't formalise testing because the result string varies by what the local timezone is. Fair enough.  But _all subsequent results are identical_ so we could be 'more formal' by creating a first 'expected result' and then comparing the output from all supported formats against it.

I think I may implement that for the next round.  Time to get the current release out the door.

**Comment by bobjansen** [comment]
Good point, didn't think about that. I was thinking that at least for TZ UTC and `asUTC=TRUE` shouldn't depend on timezone and that would be a good basis for testing the formats.

I agree about the release, no complaints left on my side :)

**Comment by eddelbuettel** [comment]
Yes, but I think I tried with UTC and win-builder still had something against it.  Not sure.  That is one option, as is the 'relative comparison' of all outputs being equivalent.   We'll get there.

I just rolled up a new release 0.1.0 and shipped it.  With some luck I get into the (European hours) evening processing window at CRAN.

Thanks a bunch for all your continued help along the way.  We are making progress here.

**Comment by eddelbuettel** [comment]
Meh. `anytime_0.1.0` fails tests for Uwe on Windows, when passed two win-builder tests from my upload.
